accepted/2022/enable-512-vectors.md

```diff
@@ -0,0 +1,518 @@
+
+### VectorMask Usage
+
+
+For each `Vector` API, we introduce a corresponding `VectorMask`, which abstracts away low-level bit-masking and instead allows to express conditional SIMD processing as boolean logic over `Vector` APIs.
+
+For example, in the following snippet `SumVector`, we perform a conditional add of `Vector` elements if the source `Vector` `v1`'s element is greater than 0 and not equal to 5:
+
+
+```C#
+public Vector<int> SumVector(ReadOnlySpan<int> source)
+{
+  Vector<int> vresult = Vector<int>.Zero;
+
+  for (int i = 0; i < source.Length;; i += Vector<int>.Count)
+  {
+    Vector<int> v1 = new Vector<int>(source.Slice(i));
+    VectorMask<int> mask = v1.GreaterThanByVectorMask(0) & v1.NotEqualsByVectorMask(5);
+    vresult = Vector.Add(vresult, v1, mask);
+  }
+
+  return vresult;
+}
+```
+
+Note that the `VectorMask` is built directly expressing the condition of the source vector `v1`. This roughly translates to a scalar loop...
+
+```C#
+for (int i = 0; i < source.Length; i++)
+{
+  if (source[i] > 0 && source[i] != 5)
+    result[i] += source[i];
+}
+```
+
+Importantly, this conditional logic is not limited to comparing `Vector` against constants. In the following example, we perform a conditional add if the elements of `v1` are less than `v2`, i.e., `v2`:
+
+```C#
+public Vector256<int> SumTwoVector256(ReadOnlySpan<int> s1, ReadOnlySpan<int> s2)
+{
+  Vector256<int> vresult = Vector<int>.Zero;
+
+  for (int i = 0; i < s1.Length;; i += Vector256<int>.Count)
+  {
+    Vector256<int> v1 = new Vector256<int>(s1.Slice(i));
+    Vector256<int> v2 = new Vector256<int>(s2.Slice(i));
+
+    VectorMask256<int> mask = v1.LessThanByVectorMask(v2);
+    
+    vresult = Vector256.Add(vresult, v1, mask);
+  }
+
+  return vresult;
+}
+```
+
+which roughly translates to the following scalar loop logic...
+
+```C#
+for (int i = 0; i < s1.Length; i++)
+{
+  if (s1[i] < 0 && s2[i])
+    result[i] += s1[i];
+}
+```
+
+As `VectorMask256` expresses that we have a masked condition for `Vector256`, so to will we have `VectorMask128`, and `VectorMask64`. In addition, when using the variable length `Vector<T>` API, we have `VectorMask<T>`, as seen the in the following example:
+
+
+```C#
+public Vector<int> SumVector(ReadOnlySpan<int> source)
+{
+  Vector<int> vresult = Vector<int>.Zero;
+
+  for (int i = 0; i < source.Length;; i += Vector<int>.Count)
+  {
+    Vector<int> v1 = new Vector<int>(source.Slice(i));
+    VectorMask<int> mask = v1.GreaterThanByVectorMask(0) & v1.NotEqualByVectorMask(5);
+    vresult = Vector.Add(vresult, v1, mask);
+  }
+
+  return vresult;
+}
+```
+
+Where `VectorMask<T>` expresses that the number of elements the condition applies to is variable-length, and determined by the JIT at runtime (though it must be compatible with `Vector<T>` selected length).
+
+Lastly, we propose to create a `VectorMask` using builtin C# boolean expressions by passing a lambda to a special `MaskExpr` API:
```

Comment:
I think there are a few issues with `MaskExpr` related to how the JIT operates today and what optimizations it can enable.

I expect we'd have something "simpler" and easier to migrate to if we simply had a `.AsMask()` API which effectively creates a `VectorMask<T>` from the most significant bits of each element.

This would then be:
```csharp
Vector256<int> vmask = Vector256.GreaterThan(v1, Vector256<int>.Zero) & (v1 != Vector256.Create(5);
VectorMask256<int> mask = vmask.AsMask();
```

This would be fairly "natural" to translate over from:
```csharp
int mask = vmask.ExtractMostSignificantBits();
```

but would work for variable length vectors and would provide the mask as an abstraction rather than strictly as an `int`.

Reply:
It is somewhat unfortunate that `==`, `!=`, `<`, `<=`, `>`, and `>=` return `bool` rather than `Vector###<T>` and so we can't express it as `(v1 > Vector128<int>.Zero) && (v1 != Vector256.Create(5))`, but that was done to follow existing .NET guidelines/conventions and since overloading by return type isn't feasible.

Reply:
We still expect to be able to compare masks directly as well right, or do you see that going away with the `AsMask()` API, i.e., would we allow something like...

```C#
Vector256<int> vmask = Vector256.GreaterThan(v1, Vector256<int>.Zero).AsMask();
Vector256<int> vmask2 = Vector256.NotEquals(v1,  Vector256.Create(5).AsMask();
VectorMask256<int> mask = vmask & vmask2;
```

Reply:
> We still expect to be able to compare masks directly as well right

Right, I believe everything you proposed being possible with `VectorMask` will still be possible, its only a difference in how you get the mask.

Doing:
```csharp
Vector256<int> vmask = Vector256.GreaterThan(v1, Vector256<int>.Zero);
Vector256<int> vmask2 = Vector256.NotEquals(v1,  Vector256.Create(5));
VectorMask256<int> mask = (vmask & vmask2).AsMask();
```

-or- doing:
```csharp
VectorMask256<int> mask1 = Vector256.GreaterThan(v1, Vector256<int>.Zero).AsMask();
VectorMask256<int> mask2 = Vector256.NotEquals(v1,  Vector256.Create(5).AsMask();
VectorMask256<int> mask = mask1 & mask2;
```

should be basically identical, the only difference is when you create the `VectorMask256` type. I'd expect them to be "equally performant" on AVX-512. I'd expect the former to be "more performant" on AVX2 and prior.